substreams/crates/tycho-substreams/src/block_storage.rs

```diff
@@ -0,0 +1,79 @@
+use std::collections::HashMap;
+
+use substreams_ethereum::pb::{eth, eth::v2::block::DetailLevel};
+
+use crate::{
+    models::{ContractSlot, StorageChanges, Transaction},
+    pb::tycho::evm::v1::TransactionStorageChanges,
+};
+
+#[allow(dead_code)]
+/// Helper function to extract all storage changes on a block.
+/// The raw block information collected is intended to be used by the DCI (Dynamic Contract Indexer)
+/// to extract and index relevant changes. This is specifically for dynamically identified contracts
+/// that the DCI has chosen to index. Note that core protocol data should still be properly
+/// integrated and indexed by the substreams package as per usual.
+///
+/// ## Panics
+/// Panics if the provided block is not an extended block model, as indicated by its detail level.
+///
+/// ## Warning
+/// ⚠️ This function *only* works if the **extended block model** is available,
+/// more [here](https://streamingfastio.medium.com/new-block-model-to-accelerate-chain-integration-9f65126e5425)
+fn get_block_storage_changes(block: &eth::v2::Block) -> Vec<TransactionStorageChanges> {
+    if block.detail_level != Into::<i32>::into(DetailLevel::DetaillevelExtended) {
+        panic!("Only extended blocks are supported");
+    }
+    let mut block_storage_changes = vec![];
+
+    for block_tx in block.transactions() {
+        let transaction: Transaction = block_tx.into();
+
+        let mut storage_changes = vec![];
+        for call in block_tx.calls.iter() {
+            // Filter out calls that are reverted
+            if !call.state_reverted {
+                for storage_change in call.storage_changes.iter() {
+                    storage_changes.push(storage_change.clone());
+                }
```

Comment:
More efficient than pushing to a vec one by one is just to extend the vec:
```suggestion
                storage_changes.extend(call.storage_changes.iter().cloned());
```

Reply:
Update: actually I'd collect this data grouped by address already. Then when we sort by ordinal we can sort by the changes per address (much more efficient as we're sorting smaller chunks of data). So here I would collect a `let mut changes_by_address = HashMap::new();`
```suggestion
                for storage_change in call.storage_changes.iter() {
                    changes_by_address
                        .entry(change.address.clone())
                        .or_default()
                        .push(change.clone());
                }
```
Then later instead of iterating multiple times through these maps and vecs, you could just do it once:
```
        // For each address, sort by ordinal and collect latest changes per slot
        let storage_changes: Vec<StorageChanges> = changes_by_address
            .into_iter()
            .map(|(address, mut changes)| {
                changes.sort_unstable_by_key(|change| change.ordinal);

                // Collect latest change per slot
                let mut latest_changes: HashMap<Vec<u8>, ContractSlot> = HashMap::new();
                for change in changes {
                    latest_changes.insert(
                        change.key.clone(),
                        ContractSlot { slot: change.key, value: change.new_value },
                    );
                }

                StorageChanges { address, slots: latest_changes.into_values().collect() }
            })
            .collect();
```